# Advanced Dataframes Exercises
This notebook includes exercises part I, part II and part III

<hr style="border:2px solid gray">

# Advanced Dataframes Exercises: Part I

Create a notebook or python script ```named advanced_dataframes``` to do your work in for these exercises.

1. Run ```python -m pip install pymysql``` from your terminal to install the mysql client (any folder is fine)

2. cd into your exercises folder for this module and run echo ```env.py >> .gitignore```

3. Create a function named ```get_db_url```. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

4. Use your function to obtain a connection to the ```employees``` database.

5. Once you have successfully run a query:
    - a. Intentionally make a typo in the database url. What kind of error message do you see?
    - b. Intentionally make an error in your SQL query. What does the error message look like?

6. Read the ```employees``` and ```titles``` tables into two separate DataFrames.

7. How many rows and columns do you have in each DataFrame? Is that what you expected?

8. Display the summary statistics for each DataFrame.

9. How many unique titles are in the ```titles``` DataFrame?

10. What is the oldest date in the ```to_date``` column?

11. What is the most recent date in the ```to_date``` column?

<hr style="border:0.5px solid black">
<hr style="border:0.5px solid black">

In [10]:
from env import get_connection
import pandas as pd
import numpy as np

from sqlalchemy import create_engine, text

<b>3. Create a function named ```get_db_url```</b>

It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [11]:
def get_db_url(database):
    return get_connection(database)
    #return f'mysql+pymysql://{user}:{password}@{host}/{database}'

<hr style="border:0.5px solid grey">

<b>4. Use your function to obtain a connection to the ```employees``` database.

In [12]:
# SQL database and query
url = get_db_url('employees')
query = text('SELECT * FROM employees')

In [13]:
# Acquire dataframe from database
engine = create_engine(url)
df = pd.read_sql(query, engine.connect())

In [14]:
df.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [15]:
# Info about columns, dtypes, and nulls
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


<hr style="border:0.5px solid grey">

<b>5. Once you have successfully run a query:</b>
   - a. Intentionally make a typo in the database url. What kind of error message do you see?
   - b. Intentionally make an error in your SQL query. What does the error message look like?

<b>a. Intentionally make a typo in the database url. What kind of error message do you see?</b>

In [16]:
#error in db url
url = get_db_url('emply')
engine = create_engine(url)
df = pd.read_sql(query, engine.connect())

OperationalError: (pymysql.err.OperationalError) (1049, "Unknown database 'emply'")
(Background on this error at: https://sqlalche.me/e/20/e3q8)

<b>b. Intentionally make an error in your SQL query. What does the error message look like?

In [17]:
#error in query
error_query = text('SELECT ?> FROM employees')

In [18]:
url = get_db_url('employees')
engine = create_engine(url)
error_df = pd.read_sql(error_query, engine.connect())

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '?> FROM employees' at line 1")
[SQL: SELECT ?> FROM employees]
(Background on this error at: https://sqlalche.me/e/20/f405)

<hr style="border:0.5px solid grey">

<b>6. Read the ```employees``` and ```titles``` tables into two separate DataFrames.

In [19]:
#write query for employees table
employees = text('SELECT * FROM employees')

#write query for titles table
titles = text('SELECT * FROM titles')

In [20]:
#acquire the employees table from sql
employees_df = pd.read_sql(employees, engine.connect())

In [21]:
#take a look at the dataframe
employees_df.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [22]:
#what columns, dtypes and nulls do we have
employees_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


In [23]:
#acquire the titles table from sql
titles_df = pd.read_sql(titles, engine.connect())

In [24]:
#take a look at the dataframe
titles_df.head()

emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01
3   10004         Engineer  1986-12-01  1995-12-01
4   10004  Senior Engineer  1995-12-01  9999-01-01

In [25]:
#what columns, dtypes and nulls do we have
titles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


<hr style="border:0.5px solid grey">

<b>7. How many rows and columns do you have in each DataFrame? Is that what you expected?</b>
* employees dataframe: 300,024 rows and 6 columns
* titles dataframe: 443,308 rows and 4 columns

In [26]:
#how many rows, columns are in employees
employees_df.shape

(300024, 6)

In [27]:
#how many rows, columns are in titles
titles_df.shape

(443308, 4)

<hr style="border:0.5px solid grey">

<b>8. Display the summary statistics for each DataFrame.

In [28]:
# stats for employees dataframe
employees_df.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [29]:
# stats for titles dataframe
titles_df.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

<hr style="border:0.5px solid grey">

<b>9. How many unique titles are in the ```titles``` DataFrame?
* 7 unique titles

In [30]:
# Count of each title
titles_df['title'].value_counts()

Engineer              115003
Staff                 107391
Senior Engineer        97750
Senior Staff           92853
Technique Leader       15159
Assistant Engineer     15128
Manager                   24
Name: title, dtype: int64

In [31]:
# Number of unique titles
titles_df['title'].nunique()

7

In [32]:
# Unique titles array
titles_df['title'].unique()

array(['Senior Engineer', 'Staff', 'Engineer', 'Senior Staff',
       'Assistant Engineer', 'Technique Leader', 'Manager'], dtype=object)

<hr style="border:0.5px solid grey">

<b>10. What is the oldest date in the ```to_date``` column?
* 1985

In [33]:
titles_df['to_date'].min()

datetime.date(1985, 3, 1)

<hr style="border:0.5px solid grey">

<b>11. What is the most recent date in the ```to_date``` column?
* 9999

In [34]:
titles_df['to_date'].max()

datetime.date(9999, 1, 1)

<hr style="border:2px solid blue">
<hr style="border:2px solid blue">

# Advanced Dataframes Exercises: Part II

1. Copy the ```users``` and ```roles``` DataFrames from the examples above.

2. What is the result of using a ```right``` join on the DataFrames?

3. What is the result of using an ```outer``` join on the DataFrames?

4. What happens if you drop the foreign keys from the DataFrames and try to merge them?

5. Load the ```mpg``` dataset from PyDataset.

6. Output and read the documentation for the ```mpg``` dataset.

7. How many rows and columns are in the dataset?

8. Check out your column names and perform any cleanup you may want on them.

9. Display the summary statistics for the dataset.

10. How many different manufacturers are there?

11. How many different models are there?

12. Create a column named ```mileage_difference``` like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

13. Create a column named ```average_mileage``` like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

14. Create a new column on the ```mpg``` dataset named ```is_automatic``` that holds boolean values denoting whether the car has an automatic transmission.

15. Using the ```mpg``` dataset, find out which which ```manufacturer``` has the best miles per gallon on average?

16. Do automatic or manual cars have better miles per gallon?

<hr style="border:0.5px solid black">
<hr style="border:0.5px solid black">

<b>1. Copy the ```users``` and ```roles``` DataFrames from the examples above.

In [35]:
# Build users dataframe
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [36]:
# Build roles dataframe
roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

<hr style="border:0.5px solid grey">

<b>2. What is the result of using a ```right``` join on the DataFrames?

In [41]:
right_join = pd.merge(users, roles, left_on="role_id", right_on='id', how='right', indicator=True)

In [42]:
right_join

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

<hr style="border:0.5px solid grey">

<b>3. What is the result of using an ```outer``` join on the DataFrames?

In [43]:
outer_join = pd.merge(users, roles, left_on="role_id", right_on='id', how='outer', indicator=True)

In [44]:
outer_join

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

<hr style="border:0.5px solid grey">

<b>4. What happens if you drop the foreign keys from the DataFrames and try to merge them?

In [89]:
pd.merge(users, roles, how='outer', indicator=True)

id       name  role_id      _merge
0   1        bob      1.0   left_only
1   2        joe      2.0   left_only
2   3      sally      3.0   left_only
3   4       adam      3.0   left_only
4   5       jane      NaN   left_only
5   6       mike      NaN   left_only
6   1      admin      NaN  right_only
7   2     author      NaN  right_only
8   3   reviewer      NaN  right_only
9   4  commenter      NaN  right_only

<hr style="border:0.5px solid grey">

<b>5. Load the ```mpg``` dataset from PyDataset.

In [46]:
from pydataset import data

In [47]:
mpg = data('mpg')

<hr style="border:0.5px solid grey">

<b>6. Output and read the documentation for the ```mpg``` dataset.

In [48]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




<hr style="border:0.5px solid grey">

<b>7. How many rows and columns are in the dataset?
* 234 rows and 11 coluns

In [49]:
mpg.shape

(234, 11)

<hr style="border:0.5px solid grey">

<b>8. Check out your column names and perform any cleanup you may want on them.

In [50]:
# Clean column names
mpg = mpg.rename(columns={'hwy':'highway', 'cty':'city', 'displ':'displacement', 'cyl':'cylinders', 'drv': 'drive','fl':'fuel'})

In [51]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displacement  234 non-null    float64
 3   year          234 non-null    int64  
 4   cylinders     234 non-null    int64  
 5   trans         234 non-null    object 
 6   drive         234 non-null    object 
 7   city          234 non-null    int64  
 8   highway       234 non-null    int64  
 9   fuel          234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


<hr style="border:0.5px solid grey">

<b>9. Display the summary statistics for the dataset.

In [52]:
mpg.describe()

displacement         year   cylinders        city     highway
count    234.000000   234.000000  234.000000  234.000000  234.000000
mean       3.471795  2003.500000    5.888889   16.858974   23.440171
std        1.291959     4.509646    1.611534    4.255946    5.954643
min        1.600000  1999.000000    4.000000    9.000000   12.000000
25%        2.400000  1999.000000    4.000000   14.000000   18.000000
50%        3.300000  2003.500000    6.000000   17.000000   24.000000
75%        4.600000  2008.000000    8.000000   19.000000   27.000000
max        7.000000  2008.000000    8.000000   35.000000   44.000000

<hr style="border:0.5px solid grey">

<b>10. How many different manufacturers are there?
* 15 different manufacturers

In [53]:
# Unique manufactuerers
mpg['manufacturer'].nunique()

15

In [54]:
# Count of each manufactuer
mpg['manufacturer'].value_counts()

dodge         37
toyota        34
volkswagen    27
ford          25
chevrolet     19
audi          18
hyundai       14
subaru        14
nissan        13
honda          9
jeep           8
pontiac        5
land rover     4
mercury        4
lincoln        3
Name: manufacturer, dtype: int64

<hr style="border:0.5px solid grey">

<b>11. How many different ```models``` are there?
* 38 different models

In [55]:
#how many unique models
mpg['model'].nunique()

38

In [56]:
#get a count on each model
mpg['model'].value_counts()

caravan 2wd               11
ram 1500 pickup 4wd       10
civic                      9
jetta                      9
dakota pickup 4wd          9
mustang                    9
impreza awd                8
a4 quattro                 8
grand cherokee 4wd         8
a4                         7
toyota tacoma 4wd          7
camry solara               7
camry                      7
tiburon                    7
sonata                     7
passat                     7
f150 pickup 4wd            7
durango 4wd                7
explorer 4wd               6
new beetle                 6
altima                     6
4runner 4wd                6
forester awd               6
gti                        5
c1500 suburban 2wd         5
corolla                    5
corvette                   5
malibu                     5
grand prix                 5
k1500 tahoe 4wd            4
pathfinder 4wd             4
mountaineer 4wd            4
range rover                4
maxima                     3
navigator 2wd 

<hr style="border:0.5px solid grey">

<b>12. Create a column named ```mileage_difference``` like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [57]:
#new column of mileage difference
mpg['mileage_difference'] = mpg['highway'] -mpg['city']
mpg.head()

manufacturer model  displacement  year  cylinders       trans drive  city  \
1         audi    a4           1.8  1999          4    auto(l5)     f    18   
2         audi    a4           1.8  1999          4  manual(m5)     f    21   
3         audi    a4           2.0  2008          4  manual(m6)     f    20   
4         audi    a4           2.0  2008          4    auto(av)     f    21   
5         audi    a4           2.8  1999          6    auto(l5)     f    16   

   highway fuel    class  mileage_difference  
1       29    p  compact                  11  
2       29    p  compact                   8  
3       31    p  compact                  11  
4       30    p  compact                   9  
5       26    p  compact                  10

In [58]:
# Alternate version using '.assign'
mpg.assign(mileage_difference=mpg['highway'] - mpg['city'])

manufacturer   model  displacement  year  cylinders       trans drive  \
1           audi      a4           1.8  1999          4    auto(l5)     f   
2           audi      a4           1.8  1999          4  manual(m5)     f   
3           audi      a4           2.0  2008          4  manual(m6)     f   
4           audi      a4           2.0  2008          4    auto(av)     f   
5           audi      a4           2.8  1999          6    auto(l5)     f   
..           ...     ...           ...   ...        ...         ...   ...   
230   volkswagen  passat           2.0  2008          4    auto(s6)     f   
231   volkswagen  passat           2.0  2008          4  manual(m6)     f   
232   volkswagen  passat           2.8  1999          6    auto(l5)     f   
233   volkswagen  passat           2.8  1999          6  manual(m5)     f   
234   volkswagen  passat           3.6  2008          6    auto(s6)     f   

     city  highway fuel    class  mileage_difference  
1      18       29    p  compact                  11  
2      21       29    p  compact                   8  
3      20       31    p  compact                  11  
4      21       30    p  compact                   9  
5      16       26    p  compact                  10  
..    ...      ...  ...      ...                 ...  
230    19       28    p  midsize                   9  
231    21       29    p  midsize                   8  
232    16       26    p  midsize                  10  
233    18       26    p  midsize                   8  
234    17       26    p  midsize                   9  

[234 rows x 12 columns]

<hr style="border:0.5px solid grey">

<b>13. Create a column named ```average_mileage``` like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [59]:
mpg['average_mileage']= (mpg['highway'] + mpg['city'])/2

In [60]:
mpg.head()

manufacturer model  displacement  year  cylinders       trans drive  city  \
1         audi    a4           1.8  1999          4    auto(l5)     f    18   
2         audi    a4           1.8  1999          4  manual(m5)     f    21   
3         audi    a4           2.0  2008          4  manual(m6)     f    20   
4         audi    a4           2.0  2008          4    auto(av)     f    21   
5         audi    a4           2.8  1999          6    auto(l5)     f    16   

   highway fuel    class  mileage_difference  average_mileage  
1       29    p  compact                  11             23.5  
2       29    p  compact                   8             25.0  
3       31    p  compact                  11             25.5  
4       30    p  compact                   9             25.5  
5       26    p  compact                  10             21.0

<hr style="border:0.5px solid grey">

<b>14. Create a new column on the mpg dataset named ```is_automatic``` that holds boolean values denoting whether the car has an automatic transmission.

In [61]:
# There are multiple 'auto' transmissions
mpg['trans'].value_counts()

auto(l4)      83
manual(m5)    58
auto(l5)      39
manual(m6)    19
auto(s6)      16
auto(l6)       6
auto(av)       5
auto(s5)       3
auto(s4)       3
auto(l3)       2
Name: trans, dtype: int64

In [62]:
# .contains method will filter 'auto' transmissions
mpg['is_automatic'] = mpg.trans.str.contains('auto')

mpg.head()

manufacturer model  displacement  year  cylinders       trans drive  city  \
1         audi    a4           1.8  1999          4    auto(l5)     f    18   
2         audi    a4           1.8  1999          4  manual(m5)     f    21   
3         audi    a4           2.0  2008          4  manual(m6)     f    20   
4         audi    a4           2.0  2008          4    auto(av)     f    21   
5         audi    a4           2.8  1999          6    auto(l5)     f    16   

   highway fuel    class  mileage_difference  average_mileage  is_automatic  
1       29    p  compact                  11             23.5          True  
2       29    p  compact                   8             25.0         False  
3       31    p  compact                  11             25.5         False  
4       30    p  compact                   9             25.5          True  
5       26    p  compact                  10             21.0          True

<hr style="border:0.5px solid grey">

<b>15. Using the ```mpg``` dataset, find out which manufacturer has the best miles per gallon on average?
* Honda has the best mpg on average

In [63]:
# Count of each manufacturer
mpg['manufacturer'].value_counts()

dodge         37
toyota        34
volkswagen    27
ford          25
chevrolet     19
audi          18
hyundai       14
subaru        14
nissan        13
honda          9
jeep           8
pontiac        5
land rover     4
mercury        4
lincoln        3
Name: manufacturer, dtype: int64

In [64]:
# Average mileage by manufacturer
mpg.groupby('manufacturer').average_mileage.mean().sort_values(ascending = False)

manufacturer
honda         28.500000
volkswagen    25.074074
hyundai       22.750000
subaru        22.428571
audi          22.027778
toyota        21.720588
pontiac       21.700000
nissan        21.346154
chevrolet     18.447368
ford          16.680000
mercury       15.625000
jeep          15.562500
dodge         15.540541
lincoln       14.166667
land rover    14.000000
Name: average_mileage, dtype: float64

In [65]:
# Find best average mileage from manufacturers
mpg.groupby('manufacturer').average_mileage.mean().sort_values(ascending = False).head(1)

manufacturer
honda    28.5
Name: average_mileage, dtype: float64

<hr style="border:0.5px solid grey">

<b>16. Do automatic or manual cars have better miles per gallon?
*  Manual transmission gets better mpg

In [66]:
# Average mileage for auto vs manual
best_car = mpg.groupby('is_automatic').average_mileage.agg(['mean'])
best_car

mean
is_automatic           
False         22.227273
True          19.130573

<hr style="border:2px solid blue">
<hr style="border:2px solid blue">

# Advanced Dataframes Exercises: Part III

1. Use your get_db_url function to help you explore the data from the chipotle database.

2. What is the total price for each order?

3. What are the most popular 3 items?

4. Which item has produced the most revenue?

5. Join the employees and titles DataFrames together.

6. For each title, find the hire date of the employee that was hired most recently with that title.

7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [67]:

import warnings
warnings.filterwarnings("ignore")

<hr style="border:0.5px solid black">
<hr style="border:0.5px solid black">

<b>1. Use your get_db_url function to help you explore the data from the chipotle database.

In [68]:
chipotle_query = text('SELECT * FROM orders')
url = get_db_url('chipotle')
engine = create_engine(url)

In [69]:
# Acquire dataframe
chipotle_df = pd.read_sql(chipotle_query, engine.connect())

In [70]:
#take a look at the dataframe
chipotle_df.head()

id  order_id  quantity                              item_name  \
0   1         1         1           Chips and Fresh Tomato Salsa   
1   2         1         1                                   Izze   
2   3         1         1                       Nantucket Nectar   
3   4         1         1  Chips and Tomatillo-Green Chili Salsa   
4   5         2         2                           Chicken Bowl   

                                                           choice_description  \
0                                                                         nan   
1                                                                [Clementine]   
2                                                                     [Apple]   
3                                                                         nan   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   

  item_price  
0     $2.39   
1     $3.39   
2     $3.39   
3     $2.39   
4    $16.98

In [71]:
# Info about dataframe
chipotle_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   id                  4622 non-null   int64 
 1   order_id            4622 non-null   int64 
 2   quantity            4622 non-null   int64 
 3   item_name           4622 non-null   object
 4   choice_description  4622 non-null   object
 5   item_price          4622 non-null   object
dtypes: int64(3), object(3)
memory usage: 216.8+ KB


<hr style="border:0.5px solid grey">

<b>2. What is the total price for each order?

In [72]:
# Clean up price by removing '$' and making it a float dtype
chipotle_df['item_price'] = chipotle_df.item_price.str.replace('$', '').astype(float)

In [73]:
#create a column for total price
chipotle_df['total_price'] = chipotle_df['quantity'] * chipotle_df['item_price']

chipotle_df.head()

id  order_id  quantity                              item_name  \
0   1         1         1           Chips and Fresh Tomato Salsa   
1   2         1         1                                   Izze   
2   3         1         1                       Nantucket Nectar   
3   4         1         1  Chips and Tomatillo-Green Chili Salsa   
4   5         2         2                           Chicken Bowl   

                                                           choice_description  \
0                                                                         nan   
1                                                                [Clementine]   
2                                                                     [Apple]   
3                                                                         nan   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   

   item_price  total_price  
0        2.39         2.39  
1        3.39         3.39  
2        3.39         3.39  
3        2.39         2.39  
4       16.98        33.96

In [74]:
order_totals = chipotle_df.groupby('order_id').total_price.agg(['sum'])
order_totals

sum
order_id       
1         11.56
2         33.96
3         12.67
4         21.00
5         13.70
...         ...
1830      23.00
1831      12.90
1832      13.20
1833      23.50
1834      28.75

[1834 rows x 1 columns]

<hr style="border:0.5px solid grey">

<b>3. What are the most popular 3 items?

In [75]:
chipotle_df['item_name'].value_counts().head(3)

Chicken Bowl           726
Chicken Burrito        553
Chips and Guacamole    479
Name: item_name, dtype: int64

<hr style="border:0.5px solid grey">

<b>4. Which item has produced the most revenue?

In [76]:
# Sort values by sum of each item
most_rev = chipotle_df.groupby('item_name').item_price.sum().sort_values(ascending = False).head(1)
most_rev

item_name
Chicken Bowl    7342.73
Name: item_price, dtype: float64

<hr style="border:0.5px solid grey">

<b>5. Join the employees and titles DataFrames together.

In [78]:

url = get_db_url('employees')
engine = create_engine(url)

In [79]:
titles_query = text('SELECT * FROM titles')

#acquire data from sql
titles_df = pd.read_sql(titles_query, engine.connect())

In [80]:
#take a look at the titles dataframe
titles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


In [81]:
employees_query = text('SELECT * FROM employees')

employees_df = pd.read_sql(employees_query, engine.connect())

In [86]:
# Merge titles and employees using 'emp_no'
new_df = titles_df.merge(employees_df, left_on='emp_no', right_on='emp_no', indicator=True)

In [87]:
new_df.head()

emp_no            title   from_date     to_date  birth_date first_name  \
0   10001  Senior Engineer  1986-06-26  9999-01-01  1953-09-02     Georgi   
1   10002            Staff  1996-08-03  9999-01-01  1964-06-02    Bezalel   
2   10003  Senior Engineer  1995-12-03  9999-01-01  1959-12-03      Parto   
3   10004         Engineer  1986-12-01  1995-12-01  1954-05-01  Chirstian   
4   10004  Senior Engineer  1995-12-01  9999-01-01  1954-05-01  Chirstian   

  last_name gender   hire_date _merge  
0   Facello      M  1986-06-26   both  
1    Simmel      F  1985-11-21   both  
2   Bamford      M  1986-08-28   both  
3   Koblick      M  1986-12-01   both  
4   Koblick      M  1986-12-01   both

In [90]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 443308 entries, 0 to 443307
Data columns (total 10 columns):
 #   Column      Non-Null Count   Dtype   
---  ------      --------------   -----   
 0   emp_no      443308 non-null  int64   
 1   title       443308 non-null  object  
 2   from_date   443308 non-null  object  
 3   to_date     443308 non-null  object  
 4   birth_date  443308 non-null  object  
 5   first_name  443308 non-null  object  
 6   last_name   443308 non-null  object  
 7   gender      443308 non-null  object  
 8   hire_date   443308 non-null  object  
 9   _merge      443308 non-null  category
dtypes: category(1), int64(1), object(8)
memory usage: 34.2+ MB


<hr style="border:0.5px solid grey">

<b>6. For each title, find the hire date of the employee that was hired most recently with that title.

In [91]:
# Newest hire date for each job title
new_df.groupby('title').hire_date.max()

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

<hr style="border:0.5px solid grey">

<b>7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [92]:
dept_title_query = text('SELECT t.emp_no, t.title, t.from_date,t.to_date, d.dept_name FROM departments AS d JOIN dept_emp AS de USING(dept_no) JOIN titles AS t USING(emp_no);')

In [93]:
dept_titles = pd.read_sql(dept_title_query, engine.connect())

dept_titles.head()

emp_no         title   from_date     to_date         dept_name
0   10011         Staff  1990-01-22  1996-11-09  Customer Service
1   10038  Senior Staff  1996-09-20  9999-01-01  Customer Service
2   10038         Staff  1989-09-20  1996-09-20  Customer Service
3   10049  Senior Staff  2000-05-04  9999-01-01  Customer Service
4   10049         Staff  1992-05-04  2000-05-04  Customer Service

In [94]:
all_titles_crosstab = pd.crosstab(
    dept_titles.dept_name, dept_titles.title)

In [95]:
all_titles_crosstab

title               Assistant Engineer  Engineer  Manager  Senior Engineer  \
dept_name                                                                    
Customer Service                   298      2362        4             2027   
Development                       7769     58135        2            49326   
Finance                              0         0        2                0   
Human Resources                      0         0        2                0   
Marketing                            0         0        2                0   
Production                        6445     49649        4            42205   
Quality Management                1831     13852        4            11864   
Research                           378      2986        2             2570   
Sales                                0         0        2                0   

title               Senior Staff  Staff  Technique Leader  
dept_name                                                  
Customer Service           13925  16150               309  
Development                 1247   1424              7683  
Finance                    12139  13929                 0  
Human Resources            12274  14342                 0  
Marketing                  13940  16196                 0  
Production                  1270   1478              6557  
Quality Management             0      0              1795  
Research                   11637  13495               393  
Sales                      36191  41808                 0

In [96]:
current_titles = dept_titles[dept_titles.to_date == dept_titles.to_date.max()]
current_titles.head()

emp_no         title   from_date     to_date         dept_name
1    10038  Senior Staff  1996-09-20  9999-01-01  Customer Service
3    10049  Senior Staff  2000-05-04  9999-01-01  Customer Service
5    10060  Senior Staff  1996-05-28  9999-01-01  Customer Service
7    10088  Senior Staff  1993-09-02  9999-01-01  Customer Service
11   10112         Staff  1998-05-01  9999-01-01  Customer Service

In [106]:
current_titles_crosstab = pd.crosstab(current_titles.dept_name, current_titles.title)
current_titles_crosstab
#switch column/row values 
# current_titles_crosstab.T

title               Assistant Engineer  Engineer  Manager  Senior Engineer  \
dept_name                                                                    
Customer Service                    68       627        1             1790   
Development                       1833     15677        1            43364   
Finance                              0         0        1                0   
Human Resources                      0         0        1                0   
Marketing                            0         0        1                0   
Production                        1542     13325        1            37156   
Quality Management                 433      3744        1            10390   
Research                            77       830        1             2250   
Sales                                0         0        1                0   

title               Senior Staff  Staff  Technique Leader  
dept_name                                                  
Customer Service           12349   3902               241  
Development                 1085    315              6117  
Finance                    10650   3199                 0  
Human Resources            10843   3416                 0  
Marketing                  12371   3880                 0  
Production                  1123    349              5210  
Quality Management             0      0              1422  
Research                   10219   3206               321  
Sales                      32032   9967                 0